Add test for a MediaConch-challenging file

jrwdunham opened the pull request:
Adds a feature that confirms that [AM PR 995](https://github.com/artefactual/archivematica/pull/995) fixes [AM issue 966](https://github.com/artefactual/archivematica/issues/966).

The new feature file mediaconch-challenging-mkv.feature is _ad hoc_ but it confirms that AM with MediaConch 17.12 installed will not hang when attempting to perform an implementation check (validation) on an .mkv file (created from a particular .mp4 file) that _does_ cause MediaConch 16.12 to hang.

Note: the feature in mediaconch-challenging-mkv.feature implicitly requires a directory in archivematica-sampledata/ called mc-challenging-file/ that contains a single .mp4 file called lunchroom_manners_512kb.mp4 which, when normalized to .mkv, will produce a file that makes MediaConch 16.12 but not 17.12 hang when performing an implementation check. The archive containing this file can be found at https://github.com/artefactual/archivematica/issues/966. If possible, the AM-sampledata repo should be modified to include mc-verbosity-test/.

ross-spencer commented on features/core/mediaconch-verbosity.feature:
```diff
@@ -0,0 +1,33 @@
+    Given a processing configuration for MediaConch verbosity testing
+    When a transfer is initiated on directory <transfer_path>
+    And the user waits for the "Validate preservation derivatives" micro-service to complete during ingest
+    Then the "Validate preservation derivatives" micro-service output is "Completed successfully" during ingest
```
Any validation task for MKV could suffer the same outcome I believe, do we need to set examples up here so that we're passing validation on originals and derivatives?

jrwdunham replied:
There are other feature files that test the MediaConch implementation check functionality in AM with good and bad .mkv files for both originals and derivatives. See

- https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/master/features/core/ingest-mkv-conformance.feature
- https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/master/features/core/transfer-mkv-conformance.feature

ross-spencer replied:
:+1:

ross-spencer commented on features/core/mediaconch-verbosity.feature:
```diff
@@ -0,0 +1,33 @@
+    And the user waits for the "Validate preservation derivatives" micro-service to complete during ingest
+    Then the "Validate preservation derivatives" micro-service output is "Completed successfully" during ingest
+    When the user waits for the "Store AIP (review)" decision point to appear during ingest
+    Then all PREMIS implementation-check-type validation events have eventOutcome = pass
```
I think this might be exclusive of this particular issue, that is, we need a task to execute successfully, but following that, a feature is allowed to fail, or succeed so that preservation decisions can be made. So maybe we don't need to get to the AIP decision point at all?

jrwdunham replied:
In this test we know we are running the MediaConch implementation checker on a transfer containing a single .mkv file (original) that we expect should pass. So we know that all "implementation-check-type" validation events are expected to pass. Looking at mediaconch_steps.py, an "implementation-check-type" validation event is one with eventType of 'validation' and an eventDetail that starts with `program="MediaConch"`.

In order to determine that all PREMIS events of a certain type are documented as passing, we **do** need to wait for the "Store AIP (review)" decision point so that we can examine the METS file.

I'm not sure that I'm fully understanding or responding to your comment here, so please elaborate if that's the case.

ross-spencer replied:
I think this is appropriate for the rationale @jrwdunham so no worries there. 

There's some nascent thoughts I have about any of the steps in the workflow that call an external program, so, Mediaconch, JHove, FITS, I think stand out. Were their progress to stall then we could imagine acceptance of a feature running such a tool to include:

* Failures are logged and can be investigated
* (but) Transfer processing continues
* Premis events aren't recorded

I think it's beyond the scope of this. I don't know how much it makes sense yet either, nor how we'd monitor it better - if it were at all possible to do it for any generic executable we might employ...

jrwdunham replied:
@ross-spencer I think it's hard to write these features so that they have conditional forking in them. I think we have to expect AM to deterministically handle certain specific test files and test transfers in certain ways. So if we write a test/feature file that throws a transfer at AM that we expect to trigger a failure at a particular micro-service because of a particular file, then we have to write that feature file so that it finds and asserts the expected results, which, as you say, would be things like the presence or absence of log files, GUI events and PREMIS events in the METS.

Then if AM suddenly starts handling our test transfers in a different way (e.g., because Siegfried is identifying a file differently), then our tests should catch that and we'll probably want to modify the tests.

That's how I see it anyway. Not sure we disagree...

ross-spencer commented on features/steps/mediaconch_steps.py:
```diff
@@ -112,6 +112,17 @@ def step_impl(context):
+    context.execute_steps(
+        'Given a base processing configuration for MediaConch tests\n'
+        'And the processing config decision "Select file format identification'
+        ' command (Transfer)" is set to "Identify using Siegfried"\n'
```
This is good detail given that it seems SF is the most likely to identify this for now. Do you want to specify a SF signature file version too? - the identification result has the potential to change in future if we don't anchor it to a specific configuration.

jrwdunham replied:
Are you saying that the tests should have a `Given` that asserts that SiegFried is using a specific signature file version? From briefly exploring the AM GUI, I don't think this information is accessible to the user. Is it?

Maybe an equivalent check would be to assert in the test that the file is identified with the expected PRONOM id.

That said, I feel like if a future version of SF stops identifying this file as fmt/199, then the test will fail at `Then all PREMIS implementation-check-type validation events have eventOutcome = pass` because there will be 0 such validation events. So if that happens, the test will catch it.

ross-spencer replied:
That sounds good @jrwdunham. (the Pronom ID)

For SF versioning, I'll do some research and potentially raise it as a ticket outside of this issue.

ross-spencer commented on features/steps/mediaconch_steps.py:
```diff
@@ -112,6 +112,17 @@ def step_impl(context):
+        'Given a base processing configuration for MediaConch tests\n'
+        'And the processing config decision "Select file format identification'
+        ' command (Transfer)" is set to "Identify using Siegfried"\n'
+        'And the processing config decision "Normalize" is set to "Normalize'
```
See above regarding Validation in the Transfer workflow as well.

jrwdunham commented on features/core/mediaconch-verbosity.feature:
```diff
@@ -0,0 +1,33 @@
+    When a transfer is initiated on directory <transfer_path>
+    And the user waits for the "Validate preservation derivatives" micro-service to complete during ingest
+    Then the "Validate preservation derivatives" micro-service output is "Completed successfully" during ingest
+    When the user waits for the "Store AIP (review)" decision point to appear during ingest
```
Nope. That's ensured by the first step: `Given a processing configuration for MediaConch verbosity testing`.

ross-spencer replied:
Ah-ha!

jrwdunham commented:
@ross-spencer I think your general point is correct: this seems like a pretty ad hoc test. Also, it doesn't really test the verbosity of the mediaconch-related micro-services as it claims to do.

Essentially, this new feature file is testing that we can throw a particular strange .mp4 file at AM and that it will get normalized to .mkv for preservation and that the MediaConch implementation check (validation) will pass for that resulting .mkv and that validation will be recorded in the METS as a PREMIS event.

Remember that there are other features/tests in this repo that test MediaConch integration more thoroughly, which I linked to in the above comments.

I would like to get this test merged simply because it demonstrates that issue 966 is fixed and it allows us to confirm that fix quickly. If I fixed up the wording of the .feature file so it clarified the purpose of the test, would you support merging it?

ross-spencer commented:
@jrwdunham no problem. Happy to see it merged with the extra clarification. I'm a huge fan of exploring all the potential uses of our acceptance tests like this. I'll get to work seeing if we can secure that file for the sample-data repository.

jrwdunham commented:
Hey @ross-spencer, I've modified this PR so that it speaks of a "challenging MKV file/transfer" instead of talking about "verbosity". I've also added a `Then` that asserts that the original file is identified as fmt/199. Finally, I've used this to confirm that the re-based and minimally modified https://github.com/artefactual/archivematica/pull/995 still works.

Do you think this is mergeable now?

ross-spencer commented:
The changes are great @jrwdunham.

jrwdunham commented:
Oops. This should have shown as merged but I deleted the branch before pushing the rebased version of it.
